L1Trigger/TrackFindingTracklet/src/KalmanFilter.cc

```diff
@@ -0,0 +1,562 @@
+      static const double baseInvdH = pow(2., baseDiffInvdH) / baseH;
+      static const double baseInvdH2 = pow(2., baseDiffInvdH2) / baseH2;
+      static const double baseHm0 = baseH * m0_->base();
+      static const double baseHm1 = baseH * m1_->base();
+      static const double baseHv0 = baseH * v0_->base() * pow(2, baseDiffHv0);
+      static const double baseHv1 = baseH * v1_->base() * pow(2, baseDiffHv1);
+      static const double baseH2v0 = baseH2 * v0_->base() * pow(2, baseDiffH2v0);
+      static const double baseH2v1 = baseH2 * v1_->base() * pow(2, baseDiffH2v1);
+      const double dH = floor((s1->H00() - s0->H00() + 1.e-11) / baseH) * baseH;
+      const double invdH = digi(1.0 / dH, baseInvdH);
+      const double invdH2 = digi(1.0 / dH / dH, baseInvdH2);
+      const double H02 = digi(s0->H00() * s0->H00(), baseH2);
+      const double H12 = digi(s1->H00() * s1->H00(), baseH2);
+      const double H22 = digi(s0->H12() * s0->H12(), baseH2);
+      const double H32 = digi(s1->H12() * s1->H12(), baseH2);
+      const double H1m0 = digi(s1->H00() * s0->m0(), baseHm0);
+      const double H0m1 = digi(s0->H00() * s1->m0(), baseHm0);
+      const double H3m2 = digi(s1->H12() * s0->m1(), baseHm1);
+      const double H2m3 = digi(s0->H12() * s1->m1(), baseHm1);
+      const double H1v0 = digi(s1->H00() * s0->v0(), baseHv0);
+      const double H0v1 = digi(s0->H00() * s1->v0(), baseHv0);
+      const double H3v2 = digi(s1->H12() * s0->v1(), baseHv1);
+      const double H2v3 = digi(s0->H12() * s1->v1(), baseHv1);
+      const double H12v0 = digi(H12 * s0->v0(), baseH2v0);
+      const double H02v1 = digi(H02 * s1->v0(), baseH2v0);
+      const double H32v2 = digi(H32 * s0->v1(), baseH2v1);
+      const double H22v3 = digi(H22 * s1->v1(), baseH2v1);
+      const double x0 = x0_->digi((s1->m0() - s0->m0()) * invdH);
+      const double x2 = x2_->digi((s1->m1() - s0->m1()) * invdH);
+      const double x1 = x1_->digi((H1m0 - H0m1) * invdH);
+      const double x3 = x3_->digi((H3m2 - H2m3) * invdH);
+      const double C00 = C00_->digi((s1->v0() + s0->v0()) * invdH2);
+      const double C22 = C22_->digi((s1->v1() + s0->v1()) * invdH2);
+      const double C01 = C01_->digi(-(H1v0 + H0v1) * invdH2);
+      const double C23 = C23_->digi(-(H3v2 + H2v3) * invdH2);
+      const double C11 = C11_->digi((H12v0 + H02v1) * invdH2);
+      const double C33 = C33_->digi((H32v2 + H22v3) * invdH2);
+      // create updated state
+      states_.emplace_back(State(s1, {x0, x1, x2, x3, 0., 0., C00, C11, C22, C33, C01, C23}));
+      state = &states_.back();
+      x0_->updateRangeActual(x0);
+      x1_->updateRangeActual(x1);
+      x2_->updateRangeActual(x2);
+      x3_->updateRangeActual(x3);
+      C00_->updateRangeActual(C00);
+      C01_->updateRangeActual(C01);
+      C11_->updateRangeActual(C11);
+      C22_->updateRangeActual(C22);
+      C23_->updateRangeActual(C23);
+      C33_->updateRangeActual(C33);
+    }
+  }
+
+  // apply final cuts
```

Comment:
What does "final" cuts mean in this comment? Are these cuts applied at the end of the KF fit to a track, after all stubs have been added to a state. Or are they cuts applied after adding each stub?

Reply:
the former